flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java

```diff
@@ -519,24 +522,8 @@ private void dispose(boolean releaseState) {
 			try {
 				numAcknowledgedTasks = -1;
 				if (!discarded && releaseState) {
-					executor.execute(new Runnable() {
-						@Override
-						public void run() {
-
-							// discard the private states.
-							// unregistered shared states are still considered private at this point.
-							try {
-								StateUtil.bestEffortDiscardAllStateObjects(operatorStates.values());
-								targetLocation.disposeOnFailure();
-							} catch (Throwable t) {
-								LOG.warn("Could not properly dispose the private states in the pending checkpoint {} of job {}.",
-									checkpointId, jobId, t);
-							} finally {
-								operatorStates.clear();
-							}
-						}
-					});
-
+					checkpointsCleaner.asyncDiscardPrivateStatesAndCountCheckpoints(operatorStates,
```

Comment:
`CheckpointsCleaner` is just a deledate that submits tasks to the executor as before. Do you mean that the synchronisation is lost because `CheckpointsCleaner` object is created on the `buildGraph()` thread ? Sorry I don't know the threading model of Flink very well yet.

Reply:
I missed that in the previous version it also was asynchronous, so synchronization is fine. Sorry for confusion.